spec.bs

```diff
@@ -6,57 +6,882 @@ Status: CG-DRAFT
+
+  The {{SharedStorageWorkletGlobalScope}}'s [=module map=]'s [=module scripts=] should each define and {{register}} one or more {{SharedStorageOperation}}s. 
+
+  Each {{SharedStorageWorkletGlobalScope}} also has an associated <dfn for=SharedStorageWorkletGlobalScope>operation map</dfn>, which is an [=map=], initially empty, of [=strings=] (denoting operation names) to [=functions=].
+
+  Currently each {{SharedStorageOperation}} registered via {{SharedStorageWorkletGlobalScope/register()}} must be one of the following two types:
+  * {{SharedStorageRunOperation}}
+  * {{SharedStorageSelectURLOperation}}
+
+  A {{SharedStorageRunOperation}} is an {{SharedStorageOperation}} that has no JavaScript return value, and hence is especially suited for calling into the [=private aggregation=] service.
+
+  A {{SharedStorageSelectURLOperation}} is an {{SharedStorageOperation}} that takes in a [=list=] of {{SharedStorageUrlWithMetadata}}s (i.e. [=dictionaries=] containing [=strings=] representing {{URL}}s each wrapped with metadata) and returns a {{long}} integer index specifying which of these {{URL}}s should be selected.
+
+  <xmp class='idl'>
+    [Exposed=SharedStorageWorklet]
+    interface SharedStorageRunOperation : SharedStorageOperation {
+      Promise<undefined> run(object data);
+    };
+
+    [Exposed=SharedStorageWorklet]
+    interface SharedStorageSelectURLOperation : SharedStorageOperation {
+      Promise<long> run(object data,
+                        FrozenArray<SharedStorageUrlWithMetadata> urls);
+    };
+
+    dictionary SharedStorageUrlWithMetadata {
+      USVString url;
+      object reportingMetadata;
+    };
+  </xmp>
+
+  ## Registering a Worklet ## {#register-worklet}
+  <div algorithm>
+    For the <dfn method for="SharedStorageWorkletGlobalScope">register(|name|, |operationCtor|)</dfn> method, run the following steps [=in parallel=]:
+
+    1. If |name| is missing or empty, throw a {{TypeError}}.
+    1. Let |operationMap| be this {{SharedStorageWorkletGlobalScope}}'s [=SharedStorageWorkletGlobalScope/operation map=].
+    1. If |operationMap| [=map/contains=] an [=map/entry=] with [=map/key=] |name|, throw a {{TypeError}}.
+    1. If |operationCtor| is missing, throw a {{TypeError}}.
+    1. Let |isConstructor| be the result of running [=IsConstructor()=] with |operationCtor|.
+    1. If |isConstructor| is false, throw a {{TypeError}}.
+    1. Let |prototype| be the result of running |operationCtor|'s [=object/[[GetPrototypeOf]]()=] method.
+    1. If |prototype| is not an [=object=], throw a {{TypeError}}.
+    1. Let |run| be the result of running [=GetMethod()=] with |prototype| and "`run`".
+    1. If |run| is undefined, throw a {{TypeError}}.
+    1. [=map/Set=] the value of |operationMap|[|name|] to |run|.
```

Comment:
Why do these steps need to run in parallel?  It seems like all of this can run synchronously or is there something potentially long/async here?  Also, if something runs in parallel we usually return a promise so the caller can know when its complete.

Reply:
I believe I addressed this now.